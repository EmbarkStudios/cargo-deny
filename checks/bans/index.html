<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>bans - cargo-deny</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="../../favicon.svg">
        <link rel="shortcut icon" href="../../favicon.png">
        <link rel="stylesheet" href="../../css/variables.css">
        <link rel="stylesheet" href="../../css/general.css">
        <link rel="stylesheet" href="../../css/chrome.css">
        <link rel="stylesheet" href="../../css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="../../FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="../../fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="../../highlight.css">
        <link rel="stylesheet" href="../../tomorrow-night.css">
        <link rel="stylesheet" href="../../ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "../../";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded "><a href="../../index.html"><strong aria-hidden="true">1.</strong> cargo-deny</a></li><li class="chapter-item expanded "><a href="../../cli/index.html"><strong aria-hidden="true">2.</strong> Command Line Tool</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../../cli/common.html"><strong aria-hidden="true">2.1.</strong> common</a></li><li class="chapter-item expanded "><a href="../../cli/init.html"><strong aria-hidden="true">2.2.</strong> init</a></li><li class="chapter-item expanded "><a href="../../cli/check.html"><strong aria-hidden="true">2.3.</strong> check</a></li><li class="chapter-item expanded "><a href="../../cli/list.html"><strong aria-hidden="true">2.4.</strong> list</a></li></ol></li><li class="chapter-item expanded "><a href="../../checks/index.html"><strong aria-hidden="true">3.</strong> Checks</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../../checks/cfg.html"><strong aria-hidden="true">3.1.</strong> config</a></li><li class="chapter-item expanded "><a href="../../checks/advisories/index.html"><strong aria-hidden="true">3.2.</strong> advisories</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../../checks/advisories/cfg.html"><strong aria-hidden="true">3.2.1.</strong> config</a></li><li class="chapter-item expanded "><a href="../../checks/advisories/diags.html"><strong aria-hidden="true">3.2.2.</strong> diagnostics</a></li></ol></li><li class="chapter-item expanded "><a href="../../checks/bans/index.html" class="active"><strong aria-hidden="true">3.3.</strong> bans</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../../checks/bans/cfg.html"><strong aria-hidden="true">3.3.1.</strong> config</a></li><li class="chapter-item expanded "><a href="../../checks/bans/diags.html"><strong aria-hidden="true">3.3.2.</strong> diagnostics</a></li></ol></li><li class="chapter-item expanded "><a href="../../checks/licenses/index.html"><strong aria-hidden="true">3.4.</strong> licenses</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../../checks/licenses/cfg.html"><strong aria-hidden="true">3.4.1.</strong> config</a></li><li class="chapter-item expanded "><a href="../../checks/licenses/diags.html"><strong aria-hidden="true">3.4.2.</strong> diagnostics</a></li></ol></li><li class="chapter-item expanded "><a href="../../checks/sources/index.html"><strong aria-hidden="true">3.5.</strong> sources</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="../../checks/sources/cfg.html"><strong aria-hidden="true">3.5.1.</strong> config</a></li><li class="chapter-item expanded "><a href="../../checks/sources/diags.html"><strong aria-hidden="true">3.5.2.</strong> diagnostics</a></li></ol></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <!-- Track and set sidebar scroll position -->
        <script>
            var sidebarScrollbox = document.querySelector('#sidebar .sidebar-scrollbox');
            sidebarScrollbox.addEventListener('click', function(e) {
                if (e.target.tagName === 'A') {
                    sessionStorage.setItem('sidebar-scroll', sidebarScrollbox.scrollTop);
                }
            }, { passive: true });
            var sidebarScrollTop = sessionStorage.getItem('sidebar-scroll');
            sessionStorage.removeItem('sidebar-scroll');
            if (sidebarScrollTop) {
                // preserve sidebar scroll position when navigating via links within sidebar
                sidebarScrollbox.scrollTop = sidebarScrollTop;
            } else {
                // scroll sidebar to current active section when navigating via "next/previous chapter" buttons
                var activeSection = document.querySelector('#sidebar .active');
                if (activeSection) {
                    activeSection.scrollIntoView({ block: 'center' });
                }
            }
        </script>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">cargo-deny</h1>

                    <div class="right-buttons">
                        <a href="../../print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="bans"><a class="header" href="#bans">bans</a></h1>
<p>The bans check is used to deny (or allow) specific crates, as well as detect and handle multiple versions of the same crate.</p>
<pre><code class="language-bash">cargo deny check bans
</code></pre>
<h2 id="use-case---denying-specific-crates"><a class="header" href="#use-case---denying-specific-crates">Use Case - Denying specific crates</a></h2>
<p>Sometimes, certain crates just don't fit in your project, so you have to remove them. However, nothing really stops them from sneaking back in due to innocuous changes like doing a <code>cargo update</code> and getting it transitively, or even forgetting to set <code>default-features = false, features = [&quot;feature-without-the-thing&quot;]</code> when the crate is pulled in via the default features of a crate you already depend on, in your entire workspace.</p>
<p>For example, we previously depended on OpenSSL as it is the &quot;default&quot; for many crates that provide TLS. This was extremely annoying as it required us to have OpenSSL development libraries installed on Windows, for both individuals and CI. We moved all of our dependencies to use the much more streamlined <code>native-tls</code> or <code>ring</code> crates instead, and now we can make sure that OpenSSL doesn't return from the grave by accident.</p>
<h2 id="use-case---duplicate-version-detection"><a class="header" href="#use-case---duplicate-version-detection">Use Case - Duplicate version detection</a></h2>
<p>The larger your project and number of external dependencies, the likelihood that you will have multiple versions of the same crate rises. This is due to two fundamental aspects of the Rust ecosystem.</p>
<ol>
<li>Cargo's dependency resolution tries to solve all the version constraints to a crate to the same version, but is totally ok with using <a href="https://stephencoakley.com/2019/04/24/how-rust-solved-dependency-hell">multiple versions</a> if it is unable to.</li>
<li>Rust has a huge (ever growing) number of crates. Every maintainer has different amounts of time and energy they can spend on their crate, not to mention different philosophies on dependencies and how often (or not) they should be updated, so it is inevitable that crates will not always agree on which version of another crate they want to use.</li>
</ol>
<p>This tradeoff of allowing multiple version of the same crate is one of the reasons that cargo is such a pleasant experience for many people new to Rust, but as with all tradeoffs, it does come with costs.</p>
<ol>
<li>More packages must be <strong>fetched</strong>, which especially impacts CI, as well as devs.</li>
<li><strong>Compile times</strong> increase, which impacts CI and devs.</li>
<li><strong>Target directory size</strong> increases, which can impact devs, or static CI environments.</li>
<li>Final <strong>binary size</strong> will also tend to increase, which can impact users.</li>
</ol>
<p>Normally, you will not really notice that you have multiple versions of the same crate unless you constantly watch your build log, but as mentioned above, it <strong>does</strong> introduce paper cuts into your workflows.</p>
<p>The intention of duplicate detection in cargo-deny is not to &quot;correct&quot; cargo's behavior, but rather to draw your attention to duplicates so that you can make an informed decision about how to handle the situation.</p>
<ul>
<li>Maybe you want to open up a PR on a crate to use a version of the duplicate that is aligned with the rest of the ecosystem.</li>
<li>Maybe the crate has actually already been updated, but the maintainer hasn't published a new version yet, and you can ask if they can publish a new one.</li>
<li>Maybe, even though the versions are supposedly incompatible according to semver, they actually aren't, and you can temporarily patch 1 or more crates to use a different version requirement without actually change the crates' code itself.</li>
<li>Sometimes having the &quot;latest and greatest&quot; is not really that important for every version, and you can just specify a lower version in your own project that matches the transitive constraint(s).</li>
<li>And finally, you don't care about a particular case of multiple versions, so you just tell cargo-deny to ignore one or more of the specific versions, and the situation will eventually resolve itself.</li>
</ul>
<h2 id="example-output"><a class="header" href="#example-output">Example output</a></h2>
<p><img src="../../output/bans.svg" alt="bans output" /></p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../../checks/advisories/diags.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next" href="../../checks/bans/cfg.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="../../checks/advisories/diags.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next" href="../../checks/bans/cfg.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="../../elasticlunr.min.js"></script>
        <script src="../../mark.min.js"></script>
        <script src="../../searcher.js"></script>

        <script src="../../clipboard.min.js"></script>
        <script src="../../highlight.js"></script>
        <script src="../../book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
