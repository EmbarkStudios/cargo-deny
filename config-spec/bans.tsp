import "./schemd.tsp";
import "./package-spec.tsp";

model Bans {
  /**
   * Determines specific crates that are allowed. If the `allow` list has one or more entries, then
   * any crate not in that list will be denied, so use with care. Each entry uses the same
   * [PackageSpec](https://embarkstudios.github.io/cargo-deny/checks/cfg.html#package-spec)
   * as other parts of cargo-deny's configuration.
   */
  allow?: PackageSpec[] = [];
}
