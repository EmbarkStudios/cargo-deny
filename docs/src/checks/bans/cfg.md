# The `[bans]` section

Contains all of the configuration for `cargo deny check bans`

## Example Config

```ini
{{#include ../../../../tests/cfg/bans.toml}}
```

### The `multiple-versions` field (optional)

Determines what happens when multiple versions of the same crate are
encountered.

* `deny` - Will emit an error for each crate with duplicates and fail the check.
* `warn` (default) - Prints a warning for each crate with duplicates, but does 
not fail the check.
* `allow` - Ignores duplicate versions of the same crate.

### The `highlight` field (optional)

When multiple versions of the same crate are encountered and `multiple-versions`
is set to `warn` or `deny`, using the `-g <dir>` option will print out a 
[dotgraph](https://www.graphviz.org/) of each of the versions and how they were 
included into the graph. This field determines how the graph is colored to help 
you quickly spot good candidates for removal or updating.

* `lowest-version` - Highlights the path to the lowest duplicate version. 
Highlighted in ![red](https://placehold.it/15/ff0000/000000?text=+)
* `simplest-path` - Highlights the path to the duplicate version with the 
fewest number of total edges to the root of the graph, which will often be the 
best candidate for removal and/or upgrading. Highlighted in 
![blue](https://placehold.it/15/0000FF/000000?text=+).
* `all` - Highlights both the `lowest-version` and `simplest-path`. If they are 
the same, they are only highlighted in 
![red](https://placehold.it/15/ff0000/000000?text=+).

![Imgur](https://i.imgur.com/xtarzeU.png)

### Crate specifier

The `allow`, `deny`, `skip`, and `skip-tree` fields all use a crate identifier 
to specify what crate(s) they want to match against.

`{ name = "some-crate-name-here", version = "<= 0.7.0" }`

#### The `name` field

The name of the crate.

#### The `version` field (optional)

An optional version constraint specifying the range of crate versions that will 
match. Defaults to all versions (`*`).

### The `allow` and `deny` fields (optional)

Determines specific crates that are allowed or denied. If the `allow` list has 
one or more entries, then any crate not in that list will be denied, so use with
care.

### The `skip` field (optional)

When denying duplicate versions, it's often the case that there is a window of
time where you must wait for eg. PRs to be accepted and new version published,
before 1 or more duplicates are gone. The `skip` field allows you to temporarily
ignore a crate during duplicate detection so that no errors are emitted, until
it is no longer need.

It is recommended to use specific version constraints for crates in the `skip`
list, as cargo-deny will emit warnings when any entry in the `skip` list no
longer matches a crate in your graph so that you can cleanup your configuration.

### The `skip-tree` field (optional)

When dealing with duplicate versions, it's often the case that a particular 
crate acts as a nexus point for a cascade effect, by either using bleeding edge 
versions of certain crates while in alpha or beta, or on the opposite end of
the specturm, a crate is using severely outdated dependencies while much of the 
rest of the ecosystem has moved to more recent versions. In both cases, it can 
be quite tedious to explicitly `skip` each transitive dependency pulled in by 
that crate that clashes with your other dependencies, which is where `skip-tree` 
comes in.

`skip-tree` entries are similar to `skip` in that they are used to specify a 
crate name and version range that will be skipped, but they also have an 
additional `depth` field used to specify how many levels from the crate will
also be skipped. A depth of `0` would be the same as specifying the crate in
the `skip` field.

Note that by default, the `depth` is infinite.

`skip-tree` is a very big hammer at the moment, and should be used with care.
