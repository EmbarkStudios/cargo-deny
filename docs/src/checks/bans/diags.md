# Bans diagnostics

### `B001` - crate is explicitly banned

A crate which is [explicitly banned](cfg.md#the-allow-and-deny-fields-optional)
was detected.

### `B002` - crate is explicitly allowed

A crate which is [explicitly allowed](cfg.md#the-allow-and-deny-fields-optional)
was detected.

### `B003` - crate is implicitly banned

When using [`bans.allow`](cfg.md#the-allow-and-deny-fields-optional), a crate
was detected that wasn't in that list.

### `B004` - found duplicate entries for crate

One or more [duplicate versions](cfg.md#the-multiple-versions-field-optional) of
the same crate were detected.

### `B005` - crate skipped when checking for duplicates

A crate version that matched an entry in
[`bans.skip`](cfg.md#the-skip-field-optional) was encountered.

### `B006` - found wildcard dependency for crate

A crate was included via a [wildcard dependency](cfg.md#the-wildcards-field-optional)
by one or more crates.

### `B007` - skipped crate was not encountered

A crate version in [`bans.skip`](cfg.md#the-skip-field-optional) was not
encountered.

### `B008` - banned crate allowed by wrapper

A crate in `bans.deny` was allowed since it was referenced by a [`wrappers`](cfg.md#the-wrappers-field-optional) crate.

### `B009` - direct parent of banned crate was not marked as a wrapper

A crate in `bans.deny` had one or more [`wrappers`](cfg.md#the-wrappers-field-optional)crates, but a crate not in that list had a direct dependency on the banned crate.

### `B010` - skip tree root was not found in the dependency graph

A crate version in [`bans.skip-tree`](cfg.md#the-skip-tree-field-optional) was
not encountered.

### `B011` - skipping crate due to root skip

A crate was skipped from being checked as a duplicate due to being transitively
referenced by a crate version in [`bans.skip-tree`](cfg.md#the-skip-tree-field-optional).
