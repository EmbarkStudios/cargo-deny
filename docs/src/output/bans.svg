<svg fill="none" viewBox="0 0 800 908" width="800" height="908" xmlns="http://www.w3.org/2000/svg">
  <foreignObject width="100%" height="100%">
    <div xmlns="http://www.w3.org/1999/xhtml">

      <style>.term-container {
  background: #282c34;
  border-radius: 5px;
  color: #abb2bf;
  word-break: break-word;
  overflow-wrap: break-word;
  font-family: Monaco, courier, monospace;
  font-size: 12px;
  line-height: 20px;
  padding: 14px 18px;
  white-space: pre-wrap;
}

.term-fg1 { font-weight: bold; }
.term-fg3 { font-style: italic; } /* italic */
.term-fg4 { text-decoration: underline; } /* underline */
.term-fg9 { text-decoration: line-through; } /* crossed-out */

.term-fg30 { color: #666; }
.term-fg31 { color: #e06c75; } /* red */
.term-fg32 { color: #98c379; } /* green */
.term-fg33 { color: #e5c07b; } /* yellow */
.term-fg34 { color: #61afef; } /* blue */
.term-fg35 { color: #c678dd; } /* magenta */
.term-fg36 { color: #56b6c2; } /* cyan */
.term-fgx9 { color: #be5046; } /* bright red */

      </style>

      <div class="term-container">cargo-deny --manifest-path examples&#47;09_bans&#47;Cargo.toml --color always check bans`
<span class="term-fgx9 term-fg1">error[duplicate]</span><span class="term-fg1">: found 2 duplicate entries for crate &#39;autocfg&#39;</span>
  <span class="term-fg34">┌─</span> &#47;home&#47;jake&#47;code&#47;cargo-deny&#47;examples&#47;09_bans&#47;Cargo.lock:1:1
  <span class="term-fg34">│</span>
<span class="term-fg34">1</span> <span class="term-fg34">│</span> <span class="term-fg31">╭</span> <span class="term-fg31">autocfg 0.1.7 registry+https:&#47;&#47;github.com&#47;rust-lang&#47;crates.io-index</span>
<span class="term-fg34">2</span> <span class="term-fg34">│</span> <span class="term-fg31">│</span> <span class="term-fg31">autocfg 1.0.0 registry+https:&#47;&#47;github.com&#47;rust-lang&#47;crates.io-index</span>
  <span class="term-fg34">│</span> <span class="term-fg31">╰───────────────────────────────────────────────────────────────────^ lock entries</span>
  <span class="term-fg34">│</span>
  <span class="term-fg34">=</span> autocfg v0.1.7
    └── (build) openssl-sys v0.9.53
        ├── native-tls v0.2.3
        │   ├── hyper-tls v0.4.1
        │   │   └── reqwest v0.10.1
        │   │       └── bans v0.1.0
        │   └── tokio-tls v0.3.0
        │       ├── hyper-tls v0.4.1 (*)
        │       └── reqwest v0.10.1 (*)
        └── openssl v0.10.26
            └── native-tls v0.2.3 (*)
  <span class="term-fg34">=</span> autocfg v1.0.0
    └── (build) indexmap v1.3.1
        └── h2 v0.2.1
            └── hyper v0.13.1
                ├── hyper-tls v0.4.1
                │   └── reqwest v0.10.1
                │       └── bans v0.1.0
                └── reqwest v0.10.1 (*)
<span class="term-fgx9 term-fg1">error[banned]</span><span class="term-fg1">: crate &#39;openssl-sys = 0.9.53&#39; is explicitly banned</span>
   <span class="term-fg34">┌─</span> examples&#47;09_bans&#47;deny.toml:10:14
   <span class="term-fg34">│</span>
<span class="term-fg34">10</span> <span class="term-fg34">│</span>     { name = <span class="term-fg31">&quot;openssl-sys&quot;</span> },
   <span class="term-fg34">│</span>              <span class="term-fg31">^^^^^^^^^^^^^</span> <span class="term-fg31">banned here</span>
   <span class="term-fg34">│</span>
   <span class="term-fg34">=</span> openssl-sys v0.9.53
     ├── native-tls v0.2.3
     │   ├── hyper-tls v0.4.1
     │   │   └── reqwest v0.10.1
     │   │       └── bans v0.1.0
     │   └── tokio-tls v0.3.0
     │       ├── hyper-tls v0.4.1 (*)
     │       └── reqwest v0.10.1 (*)
     └── openssl v0.10.26
         └── native-tls v0.2.3 (*)
bans <span class="term-fg31">FAILED</span></div>

    </div>
  </foreignObject>
</svg>
