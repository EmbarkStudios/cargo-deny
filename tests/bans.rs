use cargo_deny::{func_name, test_utils::*};

/// Covers issue <https://github.com/EmbarkStudios/cargo-deny/issues/184>
#[test]
fn cyclic_dependencies_do_not_cause_infinite_loop() {
    gather_bans(
        func_name!(),
        KrateGather::new("cyclic_dependencies"),
        Config::default(),
    );
}

#[test]
fn allow_wrappers() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("allow_wrappers/maincrate"),
        r#"
[[deny]]
name = "dangerous-dep"
wrappers = ["safe-wrapper"]
"#,
    );

    insta::assert_json_snapshot!(diags);
}

#[test]
fn disallows_denied() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("allow_wrappers/maincrate"),
        "deny = [{name = 'dangerous-dep'}]",
    );

    insta::assert_json_snapshot!(diags);
}

/// Validates that wildcard '*' dependencies can be detected and banned
#[test]
fn deny_wildcards() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("wildcards/maincrate"),
        "wildcards = 'deny'",
    );

    insta::assert_json_snapshot!(diags);
}

/// Ensures that wildcard dependencies are still banned when
/// allow-wildcard-paths is set to true but the package is public.
#[test]
fn allow_path_wildcards_public_package() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("wildcards/allow-paths-public"),
        r#"
multiple-versions = 'allow'
wildcards = 'deny'
allow-wildcard-paths = true
"#,
    );

    insta::assert_json_snapshot!(diags);
}

/// Ensures that wildcard paths are allowed for private packages
#[test]
fn allow_path_wildcards_private_package() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("wildcards/allow-paths-private"),
        r#"
multiple-versions = 'allow'
wildcards = 'deny'
allow-wildcard-paths = true
"#,
    );

    insta::assert_json_snapshot!(diags);
}

/// Ensures that multiple versions are always deterministically sorted by
/// version number
/// See <https://github.com/EmbarkStudios/cargo-deny/issues/384>
#[test]
fn deterministic_duplicate_ordering() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("duplicates"),
        "multiple-versions = 'deny'",
    );

    insta::assert_json_snapshot!(diags);
}

/// Ensures duplicate graphs match expectations
#[test]
fn duplicate_graphs() {
    use cargo_deny::bans;

    let krates = KrateGather::new("duplicates").gather();
    let cfg = "multiple-versions = 'deny'".into();

    let dup_graphs = std::sync::Arc::new(parking_lot::Mutex::new(Vec::new()));

    let duped_graphs = dup_graphs.clone();
    gather_diagnostics::<bans::cfg::Config, _, _>(&krates, func_name!(), cfg, |ctx, cs, tx| {
        bans::check(
            ctx,
            Some(Box::new(move |dg| {
                duped_graphs.lock().push(dg);
                Ok(())
            })),
            cs,
            tx,
        );
    });

    insta::assert_debug_snapshot!(dup_graphs.lock());
}

/// Ensures that we can allow duplicates generally, but deny them for specific
/// crates
#[test]
fn deny_multiple_versions_for_specific_krates() {
    let diags = gather_bans(
        func_name!(),
        KrateGather::new("duplicates"),
        r#"
multiple-versions = 'allow'
deny = [
    { name = 'block-buffer', deny-multiple-versions = true },
    { name = 'generic-array', deny-multiple-versions = true },
]
"#,
    );

    insta::assert_json_snapshot!(diags);
}
