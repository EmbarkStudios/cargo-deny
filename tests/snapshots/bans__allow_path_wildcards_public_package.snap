---
source: tests/bans.rs
expression: diags
---
[
  {
    "fields": {
      "code": "wildcard",
      "graphs": [
        {
          "Krate": {
            "name": "wildcards-test-allow-paths-public",
            "version": "0.1.0"
          }
        }
      ],
      "labels": [
        {
          "column": 1,
          "line": 1,
          "message": "wildcard crate entry",
          "span": "wildcards-test-allow-paths-dependency = '*'"
        }
      ],
      "message": "found 1 wildcard dependency for crate 'wildcards-test-allow-paths-public'. allow-wildcard-paths is enabled, but does not apply to public crates as crates.io disallows path dependencies.",
      "severity": "error"
    },
    "type": "diagnostic"
  }
]
